### DISCUSSION TITLE
DSL for Table footer

### DISCUSSION THREAD

**Discussion by Fivell** [activeadmin/activeadmin#7861] [General]
Hello @timoschilling , @seanlinsley , Hello contributors. 
One of my customer asked me to display total sum for some of columns in table footer crossing all the pages. I looked at source code and found out there was nothing in AA working out of the box for implementing such I small feature.

So I patched some classes and forgot about it. But now I need to do it again in another project.
Screenshot to clarify
 ![footer](https://cloud.githubusercontent.com/assets/120269/6202667/93e4c2ec-b500-11e4-8a29-bb5ab132db8c.png)

How this was done.
1) I patched AA internals for supporting tfoot tag in `index_as_table``and``table_for``
source -> https://gist.github.com/Fivell/d45653841cee22a890f5
2) after this I was able to do 

``` ruby
index  has_footer: true  do
    selectable_column
    actions
    id_column
    column :balance , footer_proc: ->(collection) {
           collection.except(:limit, :offset).sum(:balance)
         }
end   
```

Problems with such hack
1) every footer ceil perform new sql query (this is bad if you have many numeric columns and want to display `sum` or `avg` or `max` of each of them), the much more better way is to calculate aggregations for each of column with 1 sql query
2) DSL is ugly, DSL should be much more cleaner for such a standard feature.

Any ideas how to improve this and make it part of AA?

**Comment by ccallebs** [comment]
If we make some assumptions about what kind of data will be available in a table footer, the DSL could be simplified drastically. For your example:

``` RUBY
column :balance, sum: true
```

We could also offer mean/median/etc.

**Comment by timoschilling** [comment]
:+1: I like the idea and will think about it

**Comment by timoschilling** [comment]
@ccallebs your idea is not bad, but it's not generic. We need to map each kind of SQL Function to a attribute.
This is more generic:

``` ruby
column :balance, footer: "SUM(balance)"
```

**Comment by Fivell** [comment]
@timoschilling ,  
`footer: "SUM(balance)"` <-- this can be not enough , we should also have possibility to use view helpers(status tags ,  money formatting  etc...), decorators , etc, in footer ceil like in simple table cell

**Comment by timoschilling** [comment]
@Fivell sure, that was only 'equal' to `column :balance, sum: true`.

**Comment by Fivell** [comment]
:+1: hope we'll add this feature soon

**Comment by dmitry** [comment]
To make it possible to use formatting in the footer I suggest adding the `:footer_formatting` attribute for column DSL.

**Comment by Fivell** [comment]
after few modifications, I fix problem with multiple queries, now I use it like 

``` ruby
index  footer_data: ->(collection){collection.except(:includes, :limit, :offset).reorder("").select("sum(amount) as total_amount").to_a.first}  do

     id_column
     column :amount, footer: -> {
           @footer_data[:total_amount]
         }
     column :account
     column :notes
     column :created_at

  end
```

**Comment by timoschilling** [comment]
`footer_data` should be a proc argument and not a global method.

I don't like the API with so many `symbol: proc{}` args.

This should work:

``` ruby
column :amount, footer: &:total_amount
```

**Comment by Fivell** [comment]
@timoschilling , this doesn't work

```
 syntax error, unexpected &
```

**Comment by Fivell** [comment]
I don't like DSL with many procs too, so maybe we might think about other way
something like 

``` ruby
footer   ->(collection) {collection.totals }, as: :table do 
   column :total, for: :amount do |row|
       number_to_currency(row.total) 
   end
end
```

then we need 
1) multiple indexes `footer  as: :table` (:blog, :grid, etc)
2) scoped_collection proc for footer `->(collection) {collection.totals }`
3) helpers can be used `number_to_currency(row.total)`
4) each column should know parent column  `for: :amount`

so DSL args 
 proc for collection, options , &block

**Comment by fjosimovski** [comment]
Hello @timoschilling . Is there any progress (plan) related to this issue?

**Comment by timoschilling** [comment]
@fwebs no

**Comment by tamvm** [comment]
Other way to customize collection and add mock object for footer row like:

```
table_for bank_clients + [OpenStruct.new(balance: bank_clients.sum(:balance))] do
```

**Comment by pundit** [comment]
Hi Authors, commentors,
@timoschilling will this feature be implemented in the future?

**Comment by timoschilling** [comment]
@pundit maybe, all depends on whether someone the time it takes to build

**Comment by victorlcampos** [comment]
@timoschilling, this is a very good idea, whats the problem with original solution? ugly dsl is better than no dsl at all. And a footer_proc is not ugly for me, it permits calculate custom things like "Weighted Average" and so one.

**Comment by qiwu7** [comment]
any progress on this?

**Comment by dhurba87** [comment]
Any progress on this?
